File: first/php/library/employe.php
<?php
class Employe extends Person {
//ATRIBUTS
    private int $valueHours;
    private int $hours;
    private int $subsidy;
    private int  $calculate;
    private array $discounts = [];
    private float $totalDiscounts;
    private array $extras = [];
    private float $totalExtras = 0;

//METHOD CONSTRUCTOR
    public function __construct(int $valueHours, int $hours) {
        parent::__construct(1023, "Juan Esteban");
        $this->valueHours = $valueHours;
        $this->hours = $hours;
        $this->setSubsidy();
        $this->calculateSalary();
    }
//RETURN VALUE
    public function getValueHours(): int {
        return $this->valueHours;
    }

//GIVE A SUBSIDY VALUE
    public function setSubsidy(): void {
        if ($this->valueHours * $this->hours < 2000000) {
            $this->subsidy = 1000000;
        } else {
            $this->subsidy = 0;
        }
    }
    
//SAVE ALL DISCOUNTS IN AN ARRAY
    public function addDiscounts(Discounts $discounts): void {
        $this->discounts[] = $discounts;
    }

//SAVE ALL EXTRAS IN AN ARRAY
    public function addExtra(Extras $extra): void {
        $this->extras[] = $extra;
    }

//CALCULALTE TOTAL DISCOUNTS
    private function setTotalDiscounts(): void {
        $netSalary = ($this->valueHours * $this->hours) + $this->subsidy;
        $this->totalDiscounts = 0;
        foreach ($this->discounts as $discounts) {
            $this->totalDiscounts += $discounts->discounts($netSalary);
        }
    }

//CALCULATE TOTAL EXTRAS
    private function setTotalExtras(): void {
        $this->totalExtras = 0;
        foreach ($this->extras as $extra) {
            $this->totalExtras += $extra->extraPay($this->valueHours); 
        }
    }

//CALCULATE TOTAL SALARY WITH DISCOUNTS AND EXTRAS
    public function calculateSalary(): int {
        $this->setTotalDiscounts();
        $this->setTotalExtras();
        
        $netSalary = ($this->valueHours * $this->hours) + $this->subsidy;
        $this->calculate = $netSalary - $this->totalDiscounts + $this->totalExtras;

        return $this->calculate;
    }

//RETURN CALCULATE
    public function getCalculate(): int {
        return $this->calculate;
    }

//RETURN SUBSIDY
    public function getSubsidy(): int {
        return $this->subsidy;
    }

//RETURN TOTAL DISCOUNTS
    public function getTotalDiscounts(): float {
        return $this->totalDiscounts;
    }

//RETURN TOTAL EXTRAS
    public function getTotalExtras(): float {
        return $this->totalExtras;
    }
}
?>

File: first/java/Employe.java

import java.util.ArrayList;
import java.util.List;

public class Employe extends Person {

    private final int valueHours;
    private final int hours;
    private int subsidy;
    private float calculate;
    private final List<Discounts> discounts = new ArrayList<>();
    private float totalDiscounts;
    private final List<Extras> extras = new ArrayList<>();
    private float totalExtras = 0;

    public Employe(int valueHours, int hours) {
        super(1023, "Juan Esteban");
        this.valueHours = valueHours;
        this.hours = hours;
        setSubsidy();
        calculateSalary();
    }

    public int getValueHours() {
        return valueHours;
    }

    public void setSubsidy() {
        if ((valueHours * hours) < 2000000) {
            subsidy = 1000000;
        } else {
            subsidy = 0;
        }
    }

    public void addDiscounts(Discounts discount) {
        discounts.add(discount);
    }

    public void addExtra(Extras extra) {
        extras.add(extra);
    }

    private void setTotalDiscounts() {
        int netSalary = (valueHours * hours) + subsidy;
        totalDiscounts = 0;
        for (Discounts discount : discounts) {
            totalDiscounts += discount.discounts(netSalary);
        }
    }

    private void setTotalExtras() {
        totalExtras = 0;
        for (Extras extra : extras) {
            totalExtras += extra.extraPay(valueHours);
        }
    }

    public int calculateSalary() {
        setTotalDiscounts();
        setTotalExtras();
        int netSalary = (valueHours * hours) + subsidy;
        calculate = netSalary - totalDiscounts + totalExtras;
        return (int) calculate;
    }

    public int getCalculate() {
        return (int) calculate;
    }

    public int getSubsidy() {
        return subsidy;
    }

    public float getTotalDiscounts() {
        return totalDiscounts;
    }

    public float getTotalExtras() {
        return totalExtras;
    }
}